server-spi/src/main/java/org/keycloak/models/utils/UserModelDelegate.java

```diff
@@ -44,16 +44,6 @@ public String getId() {
         return delegate.getId();
     }
 
-    @Override
-    public String getUsername() {
```

Comment:
The problem is that this introduces potential inconsistencies: 

I have two ways to access and/or change those attributes. If one way behaves differently than the other, this will lead to subtle and hard to debug bugs. That's why I would've deprecated and deleted those methods altogether but I can see that the effort involved is much larger as of now (for instance ftl templates rely on those methods), so we keep those methods. 

This is in general true for all subclasses of `UserModel`. For the `AbstractUserAdapterFederatedStorage` this is discussed in the other comment and there is a solution to ensure consistent behaviour as long as the implementation does not override the methods to get or set attributes. 

I didn't see any other implementation of that interface I thought was modeled for inheritance, but if it is possible to provide UserModel implementations via SPI then we definitely should think about this again.

Reply:
Replied to your other comment. Hopefully once "getUsername" and "setUsername" are not implemented on UserModel, the changes in UserModelDelegate will be possible to revert, so that it always calls the delegate instead of rely on the default implementations.

Reply:
It's possible (see the two last commits). However, please note the duplicate code I had to introduce to keep consistency!